include/llbuild/Core/AttributedKeyIDs.h

```diff
@@ -0,0 +1,118 @@
+//===- AttributedKeyIDs.h ---------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLBUILD_CORE_ATTRIBUTEDKEYIDS_H
+#define LLBUILD_CORE_ATTRIBUTEDKEYIDS_H
+
+#include "llbuild/Core/KeyID.h"
+
+#include <cstdint>
+#include <utility>
+#include <vector>
+
+namespace llbuild {
+namespace core {
+
+/// A data structure representing a set of tuples (KeyID, flag) in a somewhat
+/// compact form.
+// FIXME: At some point, figure out the optimal representation for this
+// data structure, which is likely to be a lot of the resident memory size.
+class AttributedKeyIDs {
+  // Dependencies to be kept track of.
+  std::vector<KeyID> keys;
+
+  /// Flags associated with the keys.
+  std::vector<bool> flags;
+
+public:
+
+  /// Clear the contents of the set.
+  void clear() {
+    keys.clear();
+    flags.clear();
+  }
+
+  /// Check whether the set is empty.
+  bool empty() const {
+    return keys.empty();
+  }
+
+  /// Return the size of the set.
+  size_t size() const {
+    return keys.size();
+  }
+
+  /// Change the size of the set.
+  void resize(size_t newSize) {
+    keys.resize(newSize);
+    flags.resize(newSize);
+  }
+
+  /// A return value for the subscript operator[].
+  struct KeyIDAndFlag {
+    KeyID keyID;
+    bool flag;
```

Comment:
Should we build in the expectation that this may hold more than one flag?

Reply:
I think we'd refactor this structure to be more optimized (a la delta-encoding) earlier than add another member.